File: frontend/features/products/schemas.ts
import { z } from 'zod'
import { isDecimalInput, isPercentageInput } from '@/lib/utils/form-validators'

export const productKindSchema = z.enum(['PRODUCT', 'SERVICE', 'PACKAGE', 'MEMBERSHIP', 'OTHER'])
export const productStatusSchema = z.enum(['ACTIVE', 'INACTIVE'])
export const productIvaRateSchema = z.enum(['15', '5', '0', 'EXENTO'])

export const productSchema = z.object({
  id: z.string().min(1),
  tenant_id: z.string().min(1),
  sku: z.string().min(1),
  name: z.string().min(1),
  description: z.string(),
  kind: productKindSchema,
  unit: z.string().min(1),
  unit_price: z.string().min(1),
  iva_rate: productIvaRateSchema,
  discount_percentage: z.string().nullable(),
  stock_enabled: z.boolean(),
  stock_quantity: z.string().nullable(),
  low_stock_threshold: z.string().nullable(),
  status: productStatusSchema,
  created_at: z.string().min(1),
  updated_at: z.string().min(1),
  created_by: z.string(),
  updated_by: z.string(),
  version: z.coerce.number(),
})

export const productsPageSchema = z.object({
  items: z.array(productSchema),
  next_token: z.string().nullable(),
  has_more: z.boolean(),
  total: z.number().nullable().optional(),
})

export const productFormSchema = z
  .object({
    sku: z.string().trim().min(1, 'Requerido').max(36, 'Máximo 36 caracteres'),
    name: z.string().trim().min(1, 'Requerido').max(160, 'Máximo 160 caracteres'),
    description: z.string().trim().max(500, 'Máximo 500 caracteres'),
    kind: productKindSchema,
    unit: z.string().trim().min(1, 'Requerido').max(25, 'Máximo 25 caracteres'),
    unit_price: z.string().trim().refine(isDecimalInput, 'Precio inválido'),
    iva_rate: productIvaRateSchema,
    discount_percentage: z
      .string()
      .trim()
      .refine(isPercentageInput, 'El descuento debe estar entre 0 y 100')
      .or(z.literal('')),
    stock_enabled: z.boolean(),
    stock_quantity: z.string().trim().refine(isDecimalInput, 'Stock inválido').or(z.literal('')),
    low_stock_threshold: z
      .string()
      .trim()
      .refine(isDecimalInput, 'Umbral inválido')
      .or(z.literal('')),
    status: productStatusSchema,
  })
  .strict()

const productPayloadBaseSchema = z.object({
  sku: z.string().trim().min(1).max(36),
  name: z.string().trim().min(1).max(160),
  description: z.string().trim().max(500),
  kind: productKindSchema,
  unit: z.string().trim().min(1).max(25),
  unit_price: z.string().trim().refine(isDecimalInput),
  iva_rate: productIvaRateSchema,
  discount_percentage: z.string().trim().refine(isPercentageInput).nullable().optional(),
  stock_enabled: z.boolean(),
  stock_quantity: z.string().trim().refine(isDecimalInput).nullable().optional(),
  low_stock_threshold: z.string().trim().refine(isDecimalInput).nullable().optional(),
})

export const createProductSchema = productPayloadBaseSchema
export const updateProductSchema = productPayloadBaseSchema
  .extend({ status: productStatusSchema })
  .partial()

export const discountCampaignSchema = z.object({
  active: z.boolean(),
  percentage: z.string().min(1),
  updated_at: z.string().min(1),
  updated_by: z.string(),
  version: z.coerce.number(),
})

export const updateDiscountCampaignSchema = z.object({
  active: z.boolean(),
  percentage: z.string().trim().refine(isPercentageInput, 'El porcentaje debe estar entre 0 y 100'),
})

export type ProductKind = z.infer<typeof productKindSchema>
export type ProductStatus = z.infer<typeof productStatusSchema>
export type ProductIvaRate = z.infer<typeof productIvaRateSchema>
export type Product = z.infer<typeof productSchema>
export type ProductsPage = z.infer<typeof productsPageSchema>
export type ProductFormValues = z.infer<typeof productFormSchema>
export type CreateProductInput = z.infer<typeof createProductSchema>
export type UpdateProductInput = z.infer<typeof updateProductSchema>
export type DiscountCampaign = z.infer<typeof discountCampaignSchema>
export type UpdateDiscountCampaignInput = z.infer<typeof updateDiscountCampaignSchema>

File: backend/lambdas/products/schemas.py
from __future__ import annotations

from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


class CreateProductRequest(BaseModel):
    sku: str = Field(..., min_length=1, max_length=36)
    name: str = Field(..., min_length=1, max_length=160)
    description: str = Field(default="", max_length=500)
    kind: Literal["PRODUCT", "SERVICE", "PACKAGE", "MEMBERSHIP", "OTHER"] = "PRODUCT"
    unit: str = Field(default="unit", min_length=1, max_length=25)
    unit_price: Decimal = Field(..., ge=0)
    iva_rate: Literal["15", "5", "0", "EXENTO"] = "15"
    discount_percentage: Decimal | None = Field(default=None, ge=0, le=100)
    stock_enabled: bool = False
    stock_quantity: Decimal | None = Field(default=None, ge=0)
    low_stock_threshold: Decimal | None = Field(default=None, ge=0)


class UpdateProductRequest(BaseModel):
    sku: str | None = Field(default=None, min_length=1, max_length=36)
    name: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=500)
    kind: Literal["PRODUCT", "SERVICE", "PACKAGE", "MEMBERSHIP", "OTHER"] | None = None
    unit: str | None = Field(default=None, min_length=1, max_length=25)
    unit_price: Decimal | None = Field(default=None, ge=0)
    iva_rate: Literal["15", "5", "0", "EXENTO"] | None = None
    discount_percentage: Decimal | None = Field(default=None, ge=0, le=100)
    stock_enabled: bool | None = None
    stock_quantity: Decimal | None = Field(default=None, ge=0)
    low_stock_threshold: Decimal | None = Field(default=None, ge=0)
    status: Literal["ACTIVE", "INACTIVE"] | None = None


class UpdateDiscountCampaignRequest(BaseModel):
    active: bool
    percentage: Decimal = Field(..., ge=0, le=100)
